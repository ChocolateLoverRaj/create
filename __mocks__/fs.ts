import { createFsFromVolume, Volume } from 'memfs'
import { Union } from 'unionfs'
import realFs from 'fs'

const fs = new Union()
  .use(createFsFromVolume(new Volume({

  })) as any)
  .use(realFs)

export = fs
