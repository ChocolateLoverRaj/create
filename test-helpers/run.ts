import { spawn, SpawnOptionsWithoutStdio } from 'child_process'
import concat from 'concat-stream'
import { once } from 'events'
import { removeAnsi } from 'ansi-parser'

export default async (
  args: string[],
  inputs: string[] = [],
  options?: SpawnOptionsWithoutStdio
): Promise<string> => {
  const proc = spawn('node', args, options)
  proc.stdin.setDefaultEncoding('utf-8')

  let stdoutResult = ''
  proc.stdout.pipe(concat(function (result) {
    stdoutResult = result.toString()
  }))
  let stderrResult = ''
  proc.stderr.pipe(concat(result => {
    stderrResult = result.toString()
  }))
  return new Promise((resolve, reject) => {
    (async () => {
      for (let i = 0; i < inputs.length; i++) {
        let lastStr = ''
        while (true) {
          const [data] = await once(proc.stdout, 'data') as [Buffer]
          const str = data.toString()
          // Annoying ANSI escapes:
          // https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#erase-functions
          const visibleStr = str.slice(str.lastIndexOf('\x1b[2K'))
          const visibleStrText = removeAnsi(visibleStr)
            // Ignore select inputs
            .replace('>', ' ')
            // Ignore multi select inputs
            .replace('(*)', '( )')
          const shouldAnswer = visibleStr.includes('\u00BB') && lastStr !== visibleStrText
          // console.log([visibleStrText], shouldAnswer, i, inputs.length)
          lastStr = visibleStrText
          if (shouldAnswer) break
        }
        proc.stdin.write(inputs[i])
      }
      proc.stdin.end()
      // console.log('Done with prompts')
    })().catch(reject);
    (async () => {
      const [code] = await once(proc, 'exit') as [number]
      if (code === 0) return stdoutResult
      else {
        throw new Error(
        `Process exited with code: ${code}\n\nStderr below:\n${stderrResult}`
        )
      }
    })().then(resolve, reject)
  })
}

export const DOWN = '\x1B\x5B\x42'
export const UP = '\x1B\x5B\x41'
export const ENTER = '\x0D'
