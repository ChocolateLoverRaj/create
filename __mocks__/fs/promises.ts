import { promises } from 'fs'

export = promises
