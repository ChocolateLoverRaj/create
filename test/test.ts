import { join } from 'path'
import run, { DOWN, ENTER, UP } from '../test-helpers/run'
import { readFile as readFileJson } from 'jsonfile'
import { readFile, mkdtemp, rm, readdir } from 'fs/promises'
import emptyDir from 'make-empty-dir'

const testDir = join(__dirname, '../test-tmp')
const cliFilePath = join(__dirname, '../dist/index.js')
const tmpDirPrefix = join(testDir, 'test-')

const testSimple = async (cwd: string): Promise<string[]> => {
  const output = await run([cliFilePath], [
    // Yes to published
    ENTER,
    // Package name
    'a' + ENTER,
    // It's okay if 'a' already exists
    DOWN + ENTER,
    // Just CommonJS
    DOWN + ' ' + ENTER,
    // No TypeScript
    DOWN + ENTER,
    // Written in CommonJS
    ENTER,
    // No code lint
    UP + ENTER,
    // No tests
    UP + ENTER
  ], { cwd })
  const lines = output.split('\n')
  return lines
}

beforeAll(async () => await emptyDir(testDir))
jest.setTimeout(120000)
describe('detects MIT license', () => {
  // The root dir has LICENSE file, so it should have MIT
  it('prints text', async () => {
    const cwd = await mkdtemp(tmpDirPrefix)
    const lines = await testSimple(cwd)
    expect(lines[1]).toMatchSnapshot('Detected License Text')
    await rm(cwd, { recursive: true })
  })
  it('license field in package.json', async () => {
    const cwd = await mkdtemp(tmpDirPrefix)
    await testSimple(cwd)
    const packageJson = await readFileJson(join(cwd, 'package.json'))
    expect(packageJson.license).toBe('MIT')
    await rm(cwd, { recursive: true })
  })
})

describe('standard lint', () => {
  it('standard', async () => {
    const cwd = await mkdtemp(tmpDirPrefix)
    await run([cliFilePath], [
      // Yes to published
      ENTER,
      // Package name
      'a',
      ENTER,
      // It's okay if 'a' already exists
      DOWN + ENTER,
      // Just CommonJS
      DOWN,
      ' ',
      ENTER,
      // No TypeScript
      DOWN + ENTER,
      // Written in CommonJS
      ENTER,
      // Standard code lint
      ENTER,
      // No tests
      UP + ENTER,
      // No Github Actions workflows,
      DOWN + ENTER
    ], { cwd })
    await readdir(cwd)
    const packageJson = await readFileJson(join(cwd, 'package.json'))
    // Versions will change based on most recent package versions
    expect(Object.fromEntries(Object.entries(packageJson.devDependencies).map(([k, v]) =>
      [k, '***']))).toMatchSnapshot('dev dependencies')
    // Eslint config file
    await expect(readFile(join(cwd, '.eslintrc.json'), 'utf8')).resolves
      .toMatchSnapshot('.eslintrc.json')
    await rm(cwd, { recursive: true })
  })

  describe('ts-standard', () => {
    it('without mocha', async () => {
      const cwd = await mkdtemp(tmpDirPrefix)
      await run([cliFilePath], [
        // Yes to published
        ENTER,
        // Package name
        'a',
        ENTER,
        // It's okay if 'a' already exists
        DOWN + ENTER,
        // Just CommonJS
        DOWN + ' ' + ENTER,
        // TypeScript
        ENTER,
        // Written in CommonJS
        ENTER,
        // Standard code lint
        ENTER,
        // No tests
        UP + ENTER
      ], { cwd })
      const packageJson = await readFileJson(join(cwd, 'package.json'))
      // Versions will change based on most recent package versions
      expect(Object.fromEntries(Object.entries(packageJson.devDependencies).map(([k]) =>
        [k, '***']))).toMatchSnapshot('dev dependencies')
      // Eslint config file
      await expect(readFile(join(cwd, '.eslintrc.json'), 'utf8')).resolves
        .toMatchSnapshot('.eslintrc.json')
      await rm(cwd, { recursive: true })
    })
    it('with mocha', async () => {
      const cwd = await mkdtemp(tmpDirPrefix)
      await run([cliFilePath], [
        // Yes to published
        ENTER,
        // Package name
        'a',
        ENTER,
        // It's okay if 'a' already exists
        DOWN + ENTER,
        // Just CommonJS
        DOWN + ' ' + ENTER,
        // TypeScript
        ENTER,
        // Written in CommonJS
        ENTER,
        // Standard code lint
        ENTER,
        // Mocha
        ENTER,
        // No GitHub Actions workflows
        DOWN + ENTER
      ], { cwd })
      const packageJson = await readFileJson(join(cwd, 'package.json'))
      // Versions will change based on most recent package versions
      expect(Object.fromEntries(Object.entries(packageJson.devDependencies).map(([k]) =>
        [k, '***']))).toMatchSnapshot('dev dependencies')
      // Eslint config file
      await expect(readFile(join(cwd, '.eslintrc.json'), 'utf8')).resolves
        .toMatchSnapshot('.eslintrc.json')
      // Eslint tsconfig file
      await expect(readFile(join(cwd, 'eslintTsconfig.json'), 'utf8')).resolves
        .toMatchSnapshot('eslintTsconfig.json')
      await rm(cwd, { recursive: true })
    })
  })
})
