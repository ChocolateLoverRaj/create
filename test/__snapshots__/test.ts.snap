// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`detects MIT license prints text: Detected License Text 1`] = `"Detected license: MIT"`;

exports[`standard lint standard: .eslintrc.json 1`] = `
"{
  \\"root\\": true,
  \\"extends\\": \\"standard\\",
  \\"ignorePatterns\\": \\"/dist\\"
}
"
`;

exports[`standard lint standard: dev dependencies 1`] = `
Object {
  "eslint": "***",
  "eslint-config-standard": "***",
  "eslint-plugin-import": "***",
  "eslint-plugin-node": "***",
  "eslint-plugin-promise": "***",
}
`;

exports[`standard lint ts-standard with mocha: .eslintrc.json 1`] = `
"{
  \\"root\\": true,
  \\"extends\\": \\"standard-with-typescript\\",
  \\"ignorePatterns\\": \\"/dist\\",
  \\"parserOptions\\": {
    \\"project\\": \\"./eslintTsconfig.json\\"
  }
}
"
`;

exports[`standard lint ts-standard with mocha: dev dependencies 1`] = `
Object {
  "@types/mocha": "***",
  "@typescript-eslint/eslint-plugin": "***",
  "eslint": "***",
  "eslint-config-standard-with-typescript": "***",
  "eslint-plugin-import": "***",
  "eslint-plugin-node": "***",
  "eslint-plugin-promise": "***",
  "mocha": "***",
  "typescript": "***",
}
`;

exports[`standard lint ts-standard with mocha: eslintTsconfig.json 1`] = `
"{
  \\"extends\\": \\"./tsconfig.json\\",
  \\"include\\": \\"**/*.ts\\"
}
"
`;

exports[`standard lint ts-standard without mocha: .eslintrc.json 1`] = `
"{
  \\"root\\": true,
  \\"extends\\": \\"standard-with-typescript\\",
  \\"ignorePatterns\\": \\"/dist\\",
  \\"parserOptions\\": {
    \\"project\\": \\"./eslintTsconfig.json\\"
  }
}
"
`;

exports[`standard lint ts-standard without mocha: dev dependencies 1`] = `
Object {
  "@types/mocha": "***",
  "@typescript-eslint/eslint-plugin": "***",
  "eslint": "***",
  "eslint-config-standard-with-typescript": "***",
  "eslint-plugin-import": "***",
  "eslint-plugin-node": "***",
  "eslint-plugin-promise": "***",
  "mocha": "***",
  "typescript": "***",
}
`;
